### DISCUSSION TITLE
Migrating users to Zitadel (Sha512 + Salts)

### DISCUSSION THREAD

**Discussion by karlosgm75** [zitadel/zitadel#6435] [General]
I am currently evaluating the transition to Zitadel, but I'm facing challenges in migrating users. Our current system employs Sha512 hashing and Password Salts for securely storing and verifying passwords. The verification process is as follows:

```
salted = password + user.Data.Salt

newHash = salted
sha = CreateSHA512Instance()

for i = 0 to 19
    hash = ComputeHashUsingSHA512(UTF8Bytes(newHash))
    newHash = ""
    for each byte x in hash
        newHash = newHash + FormatByteAsHex(x)

end for
```

c# code:

```
var salted = password + user.Data!.Salt;

        var newHash = salted;
        using var sha = SHA512.Create();
        for (int i = 0; i < 20; i++)
        {
            var hash = sha.ComputeHash(Encoding.UTF8.GetBytes(newHash));
            newHash = hash.Aggregate("", (current, x) => current + $"{x:x2}");
        }

        return newHash == user.Data!.PasswordHash
```

I would like to know if Zitadel already supports a hashing method similar to this one, or if I would need to implement it myself. Your insights would be greatly appreciated.

**Comment by mffap** [comment]
Hi @karlosgm75 short answer: no, SHA512 is currently not supported.

Long: 
We use the [zitadel/passwap](https://github.com/zitadel/passwap) library in zitadel/zitadel. You can find the supported algorithms in the [readme](https://github.com/zitadel/passwap#algorithms). It would be awesome, if you would contribute. Could you open an issue either on the passwap library to discuss the addition of the algorithm and a potential contribution? Thanks.
